src/enlyze/client.py

```diff
@@ -150,6 +183,88 @@ def get_variables(
             for variable in self._get_variables(machine.uuid)
         ]
 
+    def _get_paginated_timeseries(
+        self,
+        *,
+        machine_uuid: str,
+        start: datetime,
+        end: datetime,
+        variables: Sequence[str],
+        resampling_interval: Optional[int],
+    ) -> Iterator[timeseries_api_models.TimeseriesData]:
+        params: dict[str, Any] = {
+            "appliance": machine_uuid,
+            "start_datetime": start.isoformat(),
+            "end_datetime": end.isoformat(),
+            "variables": ",".join(variables),
+        }
+
+        if resampling_interval:
+            params["resampling_interval"] = resampling_interval
+
+        return self._timeseries_api_client.get_paginated(
+            "timeseries", timeseries_api_models.TimeseriesData, params=params
+        )
+
+    def _get_timeseries(
+        self,
+        start: datetime,
+        end: datetime,
+        variables: Union[
+            Sequence[user_models.Variable],
+            Mapping[user_models.Variable, user_models.ResamplingMethod],
+        ],
+        resampling_interval: Optional[int] = None,
+    ) -> Optional[user_models.TimeseriesData]:
+        try:
+            variables_sequence, variables_query_parameter_list = (
+                _get_variables_sequence_and_query_parameter_list(
+                    variables, resampling_interval
+                )
+            )
+        except ValueError as e:
+            raise EnlyzeError from e
+
+        start, end, machine_uuid = validate_timeseries_arguments(
+            start, end, variables_sequence
+        )
+
+        try:
+            chunks = chunk(
+                variables_query_parameter_list,
+                MAXIMUM_NUMBER_OF_VARIABLES_PER_TIMESERIES_REQUEST,
+            )
+        except ValueError as e:
+            raise EnlyzeError from e
+
+        chunks_pages = (
+            self._get_paginated_timeseries(
+                machine_uuid=machine_uuid,
+                start=start,
+                end=end,
+                variables=chunk,
+                resampling_interval=resampling_interval,
+            )
+            for chunk in chunks
+        )
+
+        timeseries_data_chunked = [
+            _get_timeseries_data_from_pages(pages) for pages in chunks_pages
+        ]
+        if not timeseries_data_chunked or None in timeseries_data_chunked:
+            return None
```

Comment:
Shouldn't we rather raise if `None in timeseries_data_chunked`? IIUC, this would be an unexpected situation, as the timeseries API might have returned data only for some variables but not all of them.

Reply:
I've thought about this for a while and I'm still on the fence,

> the timeseries API might have returned data only for some variables but not all of them.

How would this look like? If the API returned a non-200 status code, then an exception would've been already raised by this point by the `ApiClient`.

Come to think of it, I am not even sure why we return `None` [here](https://github.com/enlyze/enlyze-python/blob/1f5fc38597d796ee6ad1e4bfc297c4c197eb4b3a/src/enlyze/client.py#L28-L34) as opposed to raising?

Reply:
> How would this look like? If the API returned a non-200 status code, then an exception would've been already raised by this point by the ApiClient.

You're right about the API returning non-200 responses. I'm rather thinking of cases where it returns
```
  "data": {
    "columns": [],
    "records": []
  }
```

which happens if there's no data returned by the DB after the filters have been applied.

> Come to think of it, I am not even sure why we return None [here](https://github.com/enlyze/enlyze-python/blob/1f5fc38597d796ee6ad1e4bfc297c4c197eb4b3a/src/enlyze/client.py#L28-L34) as opposed to raising?

> ```py
>     try:
>         timeseries_data = next(pages)
>     except StopIteration:
>         return None
> ```

Here, we could raise. I can't really think of a situation when or how we would get any 200 response but not a single (even empty) page. 

> ```py
>     if not timeseries_data.columns:
>         return None
> ```

This rightfully happens when there is no data in the backing DB, e.g. when you request a time range prior to when the machine was set up (see above). I mean, it's still weird that the API doesn't return the columns that were requested. But I think that's due to how the service was implemented, not a deliberate design decision.